docs/references/security/rust-canister-development-security-best-practices.md

```diff
@@ -183,43 +183,157 @@ A canister could be rendered unusable so it could never be upgraded again e.g. d
+
+**Property 3**: Messages from interleaving calls have no reliable execution ordering.
+
+Let's consider the above example code again, and assume the method `example` is called twice in parallel, the resulting calls being Call 1 and Call 2. The following illustration shows two possible message orderings. On the left, the first call's messages are scheduled first, and only then the second call's messages are executed. On the right, we see another possible message scheduling, where the first message of each call are executed first. 
+
+![example_orderings](_attachments/example_orderings.png)
+
+**Property 4**: On a trap / panic, modifications to the canister state for the current message are not applied.
+
+For example, if a trap in the second message (blue box) of the above example occurs, any canister state changes for that message, even earlier in the blue box, are not applied. 
+
+### Avoid traps after await
 
 #### Security Concern
 
-Panics and traps roll back the canister state. So any state change followed by a trap or panic is of concern. This is also an important concern when inter-canister calls are made. If a panic/trap occurs after an `await` to an inter-canister call, then the state is reverted to the snapshot before the inter-canister call callback invocation (and not before the entire call!).
+Traps / panics roll back the canister state, as described in Property 4 above. So any state change followed by a trap or panic can be risky. This is also an important concern when inter-canister calls are made. If a panic/trap occurs after an `await` to an inter-canister call, then the state is reverted to the snapshot before the inter-canister call callback invocation, and not before the entire call!.
 
 This may e.g. lead to the following issues:
 
--   If state changes before an inter-canister call leave the state inconsistent and there is a panic after the inter-canister call, this results in inconsistent canister state.
+- If state changes before an inter-canister call leave the state inconsistent and there is a panic after the inter-canister call, this results in inconsistent canister state.
+
+- In particular, if e.g. part of the canister state is locked before an inter-canister call and released in the callback, the lock may never be released if the callback traps. 
 
--   In particular, if allocated resources (e.g. locks or memory) from before an inter-canister call are not released this can e.g. lead to a canister being locked forever.
+- Generally, there can be bugs when data is not persisted when the developer expected it to be.
 
--   Generally, there can be bugs when data is not persisted when the developer expected it to be.
+Note that in Rust, from Rust CDK version 0.5.1, any local variables still go out of scope on panic. This prevents some of the issues above. 
```

Comment:
How does this prevent issues?

Reply:
Operations on Rust locks (such as `.lock()` or `.write()`) generally return "guard" values, which, when dropped, release the lock. So if the cleanup call drops the local variables, including the guard values, then the lock is released, so a trap doesn't "posion" the lock.